sasmodels/models/lib/magnetic_functions.c

```diff
@@ -0,0 +1,193 @@
+static double clipp(double value, double low, double high) //from kernel_iq.c
+{
+  return (value < low ? low : (value > high ? high : value));
+} 
+
+static double length(double x, double y) 
+{
+  return sqrt(x*x + y*y);
+} 
+
+static double fq_core_shell(double q, double sld_core, double radius,
+   double sld_solvent, double fp_n, double sld[], double thickness[])
+{
+  const int n = (int)(fp_n+0.5);
+  double f, r, last_sld;
+  r = radius;
+  last_sld = sld_core;
+  f = 0.;
+  for (int i=0; i<n; i++) {
+    f += M_4PI_3 * cube(r) * (sld[i] - last_sld) * sas_3j1x_x(q*r);
+    last_sld = sld[i];
+    r += thickness[i];
+  }
+  f += M_4PI_3 * cube(r) * (sld_solvent - last_sld) * sas_3j1x_x(q*r);
+  return f;
+}
+
```

Comment:
`langevin()` is never used?

Anyway the approximation is pretty bad for small q. [Wikipedia(https://en.wikipedia.org/wiki/Brillouin_and_Langevin_functions#Langevin_function) gives a Lambert's continued fraction which is good for about 1e-7 single and 1e-14 double using x / (3 + x² / (5 + x² / (7 + x²/9))), with a cutoff of 0.1. You would need a separate expansion within [0.1, 100] to get machine precision. See `explore/precision.py` for details.

Reply:
The approximation only applies at values that are 10-5 times smaller than thermal energy, otherwise the exact result is taken. Should be alright for all practical purposes.